matrix/tests_matrix.py

```diff
@@ -0,0 +1,190 @@
+
+    def test_unity(self):
+        self.assertEqual(
+            Matrix.unity(2), Matrix((1, 0), (0, 1)))
+
+        self.assertEqual(
+            Matrix.unity(3), Matrix((1, 0, 0), (0, 1, 0), (0, 0, 1)))
+
+    def test_ones(self):
+        self.assertEqual(
+            Matrix.ones(2), Matrix((1, 1), (1, 1)))
+
+        self.assertEqual(
+            Matrix.ones(3), Matrix((1, 1, 1), (1, 1, 1), (1, 1, 1)))
+
+    def test_add(self):
+        m1 = Matrix.ones(3)
+        m2 = Matrix.ones(3)
+        m3 = Matrix.ones(2)
+
+        self.assertEqual(m1 + m2, Matrix((2, 2, 2), (2, 2, 2), (2, 2, 2)))
+
+        with self.assertRaises(ValueError):
+            _ = m1 + m3
+
+        with self.assertRaises(ValueError):
+            _ = m1 + 10
+
+        with self.assertRaises(ValueError):
+            _ = m1 + 0.5
+
+        with self.assertRaises(ValueError):
+            _ = m1 - "0.5"
+
+    def test_sub(self):
+        m1 = Matrix((1, 2), (3, 4))
+        m2 = Matrix((5, 6), (7, 8))
+        m3 = Matrix.ones(3)
+
+        self.assertEqual(m1 - m2, Matrix((-4, -4), (-4, -4)))
+
+        with self.assertRaises(ValueError):
+            _ = m1 - m3
+
+        with self.assertRaises(ValueError):
+            _ = m1 - 10
+
+        with self.assertRaises(ValueError):
+            _ = m1 - 0.5
+
+        with self.assertRaises(ValueError):
+            _ = m1 - "0.5"
+
+    def test_mul(self):
+        m1 = Matrix((1, 2), (3, 4))
+        m2 = Matrix((5, 6), (7, 8))
+        m3 = Matrix.ones(4)
+
+        self.assertEqual(m1 * m2, Matrix((5, 12), (21, 32)))
+        self.assertEqual(m1 * 10, Matrix((10, 20), (30, 40)))
+        self.assertEqual(m1 * 0.5, Matrix((0.5, 1), (1.5, 2)))
+
+        # reverse mul
+        self.assertEqual(10 * m1, Matrix((10, 20), (30, 40)))
+        self.assertEqual(0.5 * m1, Matrix((0.5, 1), (1.5, 2)))
+
+        with self.assertRaises(ValueError):
+            _ = m1 * m3
+
+        with self.assertRaises(ValueError):
+            _ = m1 * "hi"
+
+        with self.assertRaises(ValueError):
+            _ = m1 * [1, 2, 3]
+
+        with self.assertRaises(ValueError):
+            _ = m1 * ["hi", "i'm", "a test"]
+
+        with self.assertRaises(ValueError):
+            _ = m1 * (1, 2)
+
+        with self.assertRaises(ValueError):
+            _ = m1 * ("hi", "i'm", "a test")
+
+    def test_div(self):
+        m1 = Matrix((1, 2), (3, 4))
+        m2 = Matrix((5, 6), (7, 8))
+        m3 = Matrix.ones(4)
+
+        self.assertEqual(m1 / m2, Matrix((0.2, 2 / 6), (3 / 7, 0.5)))
+        self.assertEqual(m1 / 10, Matrix((0.1, 0.2), (0.3, 0.4)))
+        self.assertEqual(m1 / 0.5, Matrix((2, 4), (6, 8)))
+
+        with self.assertRaises(ValueError):
+            _ = m1 / m3
+
+        with self.assertRaises(ValueError):
+            _ = m1 / "hi"
+
+        with self.assertRaises(ValueError):
+            _ = m1 / [1, 2, 3]
+
+        with self.assertRaises(ValueError):
+            _ = m1 / ["hi", "i'm", "a test"]
```

Comment:
Lots of those tests are redundant and have no actual value
You don't need to check your function against exotic types